### DISCUSSION TITLE
How to extract PID

### DISCUSSION THREAD

**Discussion by alberk8** [nHapiNET/nHapi#578] [Q&A] [answered]
I have use the parser to get the MSH which is good but to get to the PID I need to cast the hl7 to a ORU_R01.   Is it possible just to extract the PID segment without having to process the TriggerEvent?.

```
 PipeParser parser = new PipeParser();
 var hl7 = parser.Parse(sMessage);
         
 var msh = (MSH)hl7.GetStructure("MSH");
```

```
MSH|^~\&||PK5293919411E|||20240719014709.1412+0800||ORU^R01|202407190144709.14120059|P|2.6
PID|||PA002||||19941203|M||||||||||M33
PV1||I|^^&1|0
OBR||||2^DATA^99CUSTOM|||202407190144709
```

**Comment by jakubsuchybio** [comment]
```csharp
var hl7 = parser.Parse(sMessage);
var oru = (ORU_R01)hl7;
var patientResult = message.GetPATIENT_RESULT();
var pid = patientResult.PATIENT.PID;
```

But I strongly suggest to:
1. Firstly get the MSH, figure out which message type it is
2. Then get from the MSH version of the message
3. Only then cast it to message type of specific version
4. Then you get the PID from the message

**Reply by alberk8** [reply]
I am trying just to get the PID for logging without processing if the message is an ORU, ADT or any of the TriggerEvents.   Is it possible wit h the current implementation?  Thanks.

**Reply by jakubsuchybio** [reply]
There is no nice way to go around parsing a few things as I said.
But there are some ugly methods you can use, for example `dynamic`
```csharp
dynamic patientData = hl7.GetStructure("PATIENT");
dynamic pid = patientData.PID;
```

**Reply by alberk8** [reply]
Thanks,  dynamic is indeed ugly.

**Reply by milkshakeuk** [reply, marked as answer]
you could also try the `SegmentFinder` Class.

```csharp
var hl7 = parser.Parse(message);
var finder = new SegmentFinder(hl7);

var pid = (PID)finder.FindSegment("PID", 0);
var id = pid.PatientID.IDNumber.Value;
```

**Reply by milkshakeuk** [reply]
@alberk8 did you see this?

**Reply by alberk8** [reply]
@milkshakeuk @jakubsuchybio   Thank you so much.  It works very well.

**Reply by milkshakeuk** [reply]
@alberk8 if any of these suggestions worked out please mark them as the answer.
